### PR TITLE
[ITSASU-899] Add code to populate model from pre-pop

### PR DESCRIPTION
**Pull request by aug24**

### CODE REVIEW

File: app/controllers/usermatching/HomeController.scala [hunk 1]
```diff
@@ -81,6 +76,39 @@ class HomeController @Inject()(val auditingService: AuditingService,
...
+        Redirect(eligibilityRoutes.NotEligibleForIncomeTaxController.show())
+      case Left(_) => throw new InternalServerException(s"[HomeController] [index] Could not retrieve eligibility status")
+    }
+  }
+
+  private def prePopulate(utr: String, prepopMaybe: Option[PrePopData])
+                         (implicit hc: HeaderCarrier, request: Request[AnyContent]) = {
+    if (isEnabled(PrePopulate)) {
+      prepopMaybe match {
+        case None => Future.successful(Unit)
```
Comment (aug24) [inline app/controllers/usermatching/HomeController.scala:99]: If no prepop available, do nothing.

File: app/controllers/usermatching/HomeController.scala [hunk 2]
```diff
@@ -81,6 +76,39 @@ class HomeController @Inject()(val auditingService: AuditingService,
...
+        // Check presence of user information from ITSA data store by requesting an optional boolean of pre-pop flag.
+        case Some(prePop) => subscriptionDetailsService.fetchPrePopFlag(utr) map {
+          case None =>
+            populateSubscription(utr, prePop)
+          case _ =>
+            Future.successful(Unit)
```
Comment (aug24) [inline app/controllers/usermatching/HomeController.scala:105]: If there is a prepop flag, do nothing.

File: test/controllers/usermatching/HomeControllerSpec.scala [hunk 3]
```diff
@@ -18,18 +18,18 @@ package controllers.usermatching
 
 import agent.audit.mocks.MockAuditingService
 import config.MockConfig
-import config.featureswitch.FeatureSwitch.SPSEnabled
+import config.featureswitch.FeatureSwitch.{PrePopulate, SPSEnabled}
```
Comment (aug24) [inline test/controllers/usermatching/HomeControllerSpec.scala:21]: I suggest viewing this file without whitespace changes!  Add `?w=1` to the url.

File: test/controllers/usermatching/HomeControllerSpec.scala [hunk 4]
```diff
@@ -106,29 +116,172 @@ class HomeControllerSpec extends ControllerBaseSpec
...
+                      setupMockSubscriptionDetailsSaveFunctions()
+
+                      enable(SPSEnabled)
+                      enable(PrePopulate)
+
+                      val result = await(testHomeController().index(fakeRequest))
+                      status(result) must be(Status.SEE_OTHER)
+                      redirectLocation(result).get mustBe controllers.individual.sps.routes.SPSHandoffController.redirectToSPS.url
+
+                      verifySubscriptionDetailsSave(Some(1), OverseasProperty)
```
Comment (aug24) [inline test/controllers/usermatching/HomeControllerSpec.scala:178]: Got prepop data and the feature switch is 'on', so we should get four writes for the four data items.

File: app/controllers/usermatching/HomeController.scala [hunk 5]
```diff
@@ -81,6 +76,39 @@ class HomeController @Inject()(val auditingService: AuditingService,
...
+
+  private def prePopulate(utr: String, prepopMaybe: Option[PrePopData])
+                         (implicit hc: HeaderCarrier, request: Request[AnyContent]) = {
+    if (isEnabled(PrePopulate)) {
+      prepopMaybe match {
+        case None => Future.successful(Unit)
+        // Check presence of user information from ITSA data store by requesting an optional boolean of pre-pop flag.
+        case Some(prePop) => subscriptionDetailsService.fetchPrePopFlag(utr) map {
+          case None =>
+            populateSubscription(utr, prePop)
```
Comment (AlexRimmerHMRC) [inline app/controllers/usermatching/HomeController.scala:103]:
```suggestion
            populateSubscription(reference, prePop)
```
shouldn't be using utr to do this.
use the following to get the reference to save data against
```
withReference { reference =>
   ...
}
```

File: app/controllers/usermatching/HomeController.scala [hunk 6]
```diff
@@ -81,6 +76,39 @@ class HomeController @Inject()(val auditingService: AuditingService,
...
+    }
+  }
+
+  private def prePopulate(utr: String, prepopMaybe: Option[PrePopData])
+                         (implicit hc: HeaderCarrier, request: Request[AnyContent]) = {
+    if (isEnabled(PrePopulate)) {
+      prepopMaybe match {
+        case None => Future.successful(Unit)
+        // Check presence of user information from ITSA data store by requesting an optional boolean of pre-pop flag.
+        case Some(prePop) => subscriptionDetailsService.fetchPrePopFlag(utr) map {
```
Comment (AlexRimmerHMRC) [inline app/controllers/usermatching/HomeController.scala:101]:
```suggestion
        case Some(prePop) => subscriptionDetailsService.fetchPrePopFlag(reference) map {
```
use withReference to get the reference value

File: app/controllers/usermatching/HomeController.scala [hunk 7]
```diff
@@ -105,6 +133,38 @@ class HomeController @Inject()(val auditingService: AuditingService,
     }
   }
 
+  private def populateSubscription(reference: String, prePopData: PrePopData)(implicit request: Request[AnyContent]): Future[Unit] = {
+    // Set up futures so that they parallelise.
+    val futureSaveSelfEmployments = prePopData.selfEmployments
+      .map(sse => sse.map(
+        se => SelfEmploymentData(
+          UUID.randomUUID().toString,
+          None,
```
Comment (AlexRimmerHMRC) [inline app/controllers/usermatching/HomeController.scala:142]:
```suggestion
          se.businessStartDate,
```
Why is the business start date not set here?

File: app/controllers/usermatching/HomeController.scala [hunk 8]
```diff
@@ -105,6 +133,38 @@ class HomeController @Inject()(val auditingService: AuditingService,
...
+    val futureSaveUkPropertyInfo = prePopData.ukProperty
+      .map(up => PropertyModel(up.ukPropertyAccountingMethod, up.ukPropertyStartDate))
+      .map(up => subscriptionDetailsService.saveProperty(reference, up))
+      .getOrElse(Future.successful(Unit))
+    val futureSaveOverseasPropertyInfo = prePopData.overseasProperty
+      .map(op => OverseasPropertyModel(op.overseasPropertyAccountingMethod, op.overseasPropertyStartDate))
+      .map(op => subscriptionDetailsService.saveOverseasProperty(reference, op))
+      .getOrElse(Future.successful(Unit))
+    val futureSavePrePopFlag = subscriptionDetailsService.savePrePopFlag(reference, prepop = true)
+    // Wait for futures
```
Comment (AlexRimmerHMRC) [inline app/controllers/usermatching/HomeController.scala:159]: Missing one save, sorry forgot to mention it.
The business accounting method is saved separately from the rest of the self-employment data as there is only one accounting method for ALL self employments.
As each returned business from pre-pop contains an accounting method, for now, take the first business's accounting method. I will raise to see if we need stronger actions taken.

File: app/controllers/usermatching/HomeController.scala (hunk 5 above)
Comment (aug24) [reply app/controllers/usermatching/HomeController.scala]: Done.

File: app/controllers/usermatching/HomeController.scala (hunk 7 above)
Comment (aug24) [reply app/controllers/usermatching/HomeController.scala:136]: Missed it.  Done.

File: app/controllers/usermatching/HomeController.scala (hunk 8 above)
Comment (aug24) [reply app/controllers/usermatching/HomeController.scala:166]: Added.

File: app/controllers/usermatching/HomeController.scala (hunk 6 above)
Comment (aug24) [reply app/controllers/usermatching/HomeController.scala]: Done.

**Pull request merged**